htlcswitch/hop/iterator.go

```diff
@@ -251,53 +251,54 @@ func (p *OnionProcessor) DecodeHopIterators(id []byte,
 
 	tx := p.router.BeginTxn(id, batchSize)
 
-	for i, req := range reqs {
-		onionPkt := &onionPkts[i]
-		resp := &resps[i]
+	decode := func(seqNum uint16, onionPkt *sphinx.OnionPacket,
+		req DecodeHopIteratorRequest) lnwire.FailCode {
 
 		err := onionPkt.Decode(req.OnionReader)
 		switch err {
 		case nil:
 			// success
 
 		case sphinx.ErrInvalidOnionVersion:
-			resp.FailCode = lnwire.CodeInvalidOnionVersion
-			continue
+			return lnwire.CodeInvalidOnionVersion
 
 		case sphinx.ErrInvalidOnionKey:
-			resp.FailCode = lnwire.CodeInvalidOnionKey
-			continue
+			return lnwire.CodeInvalidOnionKey
 
 		default:
 			log.Errorf("unable to decode onion packet: %v", err)
-			resp.FailCode = lnwire.CodeInvalidOnionKey
-			continue
+			return lnwire.CodeInvalidOnionKey
 		}
 
 		err = tx.ProcessOnionPacket(
-			uint16(i), onionPkt, req.RHash, req.IncomingCltv,
+			seqNum, onionPkt, req.RHash, req.IncomingCltv,
 		)
 		switch err {
 		case nil:
 			// success
 
 		case sphinx.ErrInvalidOnionVersion:
-			resp.FailCode = lnwire.CodeInvalidOnionVersion
-			continue
+			return lnwire.CodeInvalidOnionVersion
 
 		case sphinx.ErrInvalidOnionHMAC:
-			resp.FailCode = lnwire.CodeInvalidOnionHmac
-			continue
+			return lnwire.CodeInvalidOnionHmac
 
 		case sphinx.ErrInvalidOnionKey:
-			resp.FailCode = lnwire.CodeInvalidOnionKey
-			continue
+			return lnwire.CodeInvalidOnionKey
 
 		default:
 			log.Errorf("unable to process onion packet: %v", err)
-			resp.FailCode = lnwire.CodeInvalidOnionKey
-			continue
+			return lnwire.CodeInvalidOnionKey
 		}
+
+		return lnwire.CodeNone
```

Comment:
nit: we could return this in the `nil` case now

Reply:
I didn't do this initially for consistency with the first switch. But something is to be said for using the shortest possible code path, so fixed.